lib/Sema/TypeCheckType.cpp

```diff
@@ -2345,15 +2345,23 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
     }
   }
 
-  if (hasFunctionAttr && !fnRepr) {
-    if (attrs.has(TAK_autoclosure)) {
+  if (attrs.has(TAK_autoclosure)) {
+    // If this is a situation where function type is wrapped
+    // into a number of parens, let's try to look through them,
+    // because parens are insignficiant here e.g.:
+    //
+    // let _: (@autoclosure (() -> Void)) -> Void = { _ in }
+    if (!ty->is<FunctionType>()) {
       // @autoclosure is going to be diagnosed when type of
       // the parameter is validated, because that attribute
       // applies to the declaration now.
       repr->setInvalid();
-      attrs.clearAttribute(TAK_autoclosure);
     }
 
+    attrs.clearAttribute(TAK_autoclosure);
```

Comment:
It's not clear why we need to remove `@autoclosure` attribute here. Could you explain?

Reply:
So it doesn't get included into the later check for `if (hasFunctionAttr && !fnRepr) {` because `@autoclosure` is currently a `FunctionAttr` and `fnRepr` is going to be `nullptr` for cases where function is wrapped into a paren or is a typealias...

Reply:
Could you add test cases for use-sites as well? like:
```
typealias GenericFunc<T> = () -> T
func test<Q>(gn: @autoclosure GenericFunc <Q>) {}
foo(gn: 12)
```
Should be type checked, right?

Reply:
Yes, I think this should be type-checked, I'll add the test for bit, it stems for the fact that `test` is correctly marked as invalid in this case, but it works after my changes.

Reply:
@rintaro Done!